File: March 2021/05.03.2021/vedat.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public class Pair {
        long first;
        long second;
        Pair(long first, long second) {
            this.first = first;
            this.second = second;
        }
    }
    public TreeMap<Integer, Pair> sums = new TreeMap<Integer, Pair>();
    
    public void traverse(TreeNode root, int depth) {
        if (root == null) {
            return;
        }
        long count = 1;
        long sum = root.val;
        if (sums.containsKey(depth)) {
            count += sums.get(depth).second;
            sum += sums.get(depth).first;
        }
        sums.put(depth, new Pair(sum, count));
        traverse(root.left, depth + 1);
        traverse(root.right, depth + 1);
    }
    
    public List<Double> averageOfLevels(TreeNode root) {
        ArrayList<Double> ans = new ArrayList<Double>(sums.size());
        traverse(root, 0);
        for (var x: sums.keySet()) {
            ans.add(sums.get(x).first * 1.0 / sums.get(x).second);
        }
        return ans;
    }
    
}

File: March 2021/09.03.2021/vedat.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void traverse(TreeNode* root, int curr_depth, int d, int v) {
        if (!root) {
            return;
        }
        
        if (d == curr_depth + 1) {
            auto left = root->left;
            auto right = root->right;
            root->left = new TreeNode(v);
            root->left->left = left;
            root->right = new TreeNode(v);
            root->right->right = right;
            return;
        }
        
        traverse(root->left, curr_depth + 1, d, v);
        traverse(root->right, curr_depth + 1, d, v);
    }
    TreeNode* addOneRow(TreeNode* root, int v, int d) {
        if (d == 1) {
            auto new_root = new TreeNode(v);
            new_root->left = root;
            return new_root;
        }
        traverse(root, 1, d, v);
        return root;
    }
};
